libnavigation-core/src/test/java/com/mapbox/navigation/core/CurrentIndicesSnapshotProviderTest.kt

```diff
@@ -0,0 +1,224 @@
+package com.mapbox.navigation.core
+
+import com.mapbox.navigation.base.internal.CurrentIndicesSnapshot
+import com.mapbox.navigation.base.trip.model.RouteProgress
+import io.mockk.every
+import io.mockk.mockk
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertNull
+import org.junit.Test
+
+class CurrentIndicesSnapshotProviderTest {
+
+    private val provider = CurrentIndicesSnapshotProvider()
+    private val currentLegIndex = 9
+    private val routeGeometryIndex = 44
+    private val legGeometryIndex = 33
+    private val routeProgress = mockk<RouteProgress> {
+        every { currentRouteGeometryIndex } returns routeGeometryIndex
+        every { currentLegProgress } returns mockk {
+            every { legIndex } returns currentLegIndex
+            every { geometryIndex } returns legGeometryIndex
+        }
+    }
+    private val expected = CurrentIndicesSnapshot(
+        currentLegIndex,
+        routeGeometryIndex,
+        legGeometryIndex
+    )
+
+    @Test
+    fun initialStateIsNull() {
+        assertNull(provider())
+    }
+
+    @Test
+    fun stateIsNullAfterClear() {
+        provider.onRouteProgressChanged(routeProgress)
+
+        provider.clear()
+
+        assertNull(provider())
+    }
+
+    @Test
+    fun stateAfterUpdate() = runBlocking {
+        provider.onRouteProgressChanged(routeProgress)
+
+        assertEquals(expected, provider())
+        assertEquals(expected, provider.getFilledIndicesAndFreeze())
+    }
+
+    @Test
+    fun updateDuringRetrieval() = runBlocking {
+        launch {
+            delay(100)
+            provider.onRouteProgressChanged(routeProgress)
+        }
+
+        assertEquals(expected, provider.getFilledIndicesAndFreeze())
+        assertEquals(expected, provider())
+    }
```

Comment:
```suggestion
    @Test
    fun updateDuringRetrieval() = runBlocking {
        launch {
            delay(100)
            provider.onRouteProgressChanged(routeProgress)
        }

        assertEquals(expected, provider.getFilledIndicesAndFreeze())
        assertEquals(expected, provider())
    }
    
    @Test
    fun updateDuringRetrieval_async() = coroutineRule.runBlockingTest {
        pauseDispatcher {
            launch {
                delay(100)
                provider.onRouteProgressChanged(routeProgress)
            }
            var value: CurrentIndicesSnapshot? = null
            val update = launch {
                value = provider.getFilledIndicesAndFreeze()
                throw AssertionError()
            }
            advanceTimeBy(50)
            update.cancel()
            advanceTimeBy(50)
            assertNull(value)
            assertEquals(expected, provider())
        }
    }
```
We also should test the cancellation.

Reply:
Added the test, thanks.